src/nutdumpdata.cc

```diff
@@ -0,0 +1,330 @@
+        const unsigned int loop_nb,
+        map_string_t& out)
+{
+    Argv args;
+    //to forge the command line to call the driver
+    //first the driver name
+    args.push_back(driver);
+    //ask the driver to do dump-data
+    args.push_back("-d");
+    args.push_back(std::to_string(loop_nb));
+    //set root as user to avoid any privildge issue
+    args.push_back("-u");
+    args.push_back("root");
+    //set a unique name with the driver name
+    args.push_back("-s");
+    std::string name="dumpdata";
+    name.append( std::to_string(rand() % 1000 + 1));
+    args.push_back(name);
+    //add all extra parameters
+    for (auto it: extra) {
+        args.push_back("-x");
+        std::string option=it.first+"="+it.second;
+        args.push_back(option);
+    }
+    return s_run_nut_dumpdata(
+                args,
+                loop_nb,
+                out);
+
+}
+
+
+/**
+ * \brief helper method to get data from a snmp-ups  
+ *
+ * \param[in] port      - port (ip:port)
+ * \param[in] community - community name. If empty, use "public"
+ * \param[out] out      - map of key/value returned by driver with output
+ * \return 0 if success, -1 otherwise
+ */
+int
+nut_dumpdata_snmp_ups(
+        const std::string& port,
+        const std::string& community,
+        map_string_t& out)
+{
+    map_string_t extra;
+    extra["port"]=port;
+    std::string comm;
+    comm = community;
+    if (comm.empty())
+        comm = "public";
+    extra["community"]=comm;
+    int loop=s_get_nut_dumpdata_loop();
+    return nut_dumpdata("snmp-ups",
+            extra,
+            loop,
+            out);
+}
+
+/**
+ * \brief helper method to get data from a netxml-ups driver 
+ *
+ * \param[in] port   - port (ip:port)
+ * \param[out] out   - map of key/value returned by driver with output
+ * \return 0 if success, -1 otherwise
+ */
+int
+nut_dumpdata_netxml_ups(
+        const std::string& port,
+        map_string_t& out)
+{
+    map_string_t extra;
+    extra["port"]=port;
+    int loop=s_get_nut_dumpdata_loop();
+    return nut_dumpdata("netxml-ups",
+            extra,
+            loop,
+            out);
+}
+
+/**
+ * \brief helper method to get data from a dummy driver 
+ *
+ * \param[in] device - dummy file device name (it is supposed to be in /etc/nut)
+ * \param[out] out   - map of key/value returned by driver with output
+ * \return 0 if success, -1 otherwise
+ */
+int
+nut_dumpdata_dummy_ups(
+        const std::string& device,
+        map_string_t& out)
+{
+    map_string_t extra;
+    extra["port"]=device;
+    int loop=s_get_nut_dumpdata_loop();
+    return nut_dumpdata("dummy-ups",
+            extra,
+            loop,
+            out);
+}
+
+
+//  --------------------------------------------------------------------------
+//  Self test of this class
+
+void
+nutdumpdata_test (bool verbose)
+{
+    printf (" * nutdumpdata ... \n");
+    //drop and prepare a dummy dev file 
+    std::remove("/etc/nut/test.dev");
```

Comment:
You do not have access to this from a testing/building account. Use the `SELFTEST_DIR_RW` as exemplified in other components.